File: modules/shared/py_packages/dw/state_machine/base_state_machine.py
"""
Game State Machine Module

This module contains the BaseStateMachine class which implements a state machine using the 
'transitions' package to control game flow.
"""

import logging
import time
# from abc import ABC
from transitions import Machine
from transitions.extensions import GraphMachine
# from .base_state import BaseState
from .pong_api import PongAPI

# class BaseStateMachine(ABC):
class BaseStateMachine():
    """
    BaseStateMachine Class

    This class implements a state machine to control game flow.
    """

    states = ['stopped', 'idle', 'intro', 'moveLeftIntro', 'moveRightIntro', 'level1', 'level2', 'level3', 'outro']
    transitions = [
        {'trigger': 'start', 'source': 'stopped', 'dest': 'idle'},
        {'trigger': 'player_ready', 'source': 'idle', 'dest': 'intro'},
        {'trigger': 'start_game', 'source': 'intro', 'dest': 'level1'},
        {'trigger': 'player_exit', 'source': 'intro', 'dest': 'idle'},
        {'trigger': 'intro_complete', 'source': 'intro', 'dest': 'move_left_intro'},
        {'trigger': 'player_exit', 'source': 'move_left_intro', 'dest': 'idle'},
        {'trigger': 'move_left_intro_complete', 'source': 'move_left_intro', 'dest': 'move_right_intro'},
        {'trigger': 'player_exit', 'source': 'move_right_intro', 'dest': 'idle'},
        {'trigger': 'start_game', 'source': 'move_right_intro', 'dest': 'level1'},
        {'trigger': 'player_exit', 'source': 'level1', 'dest': 'idle'},
        {'trigger': 'level1_complete', 'source': 'level1', 'dest': 'level2'},
        {'trigger': 'player_exit', 'source': 'level2', 'dest': 'idle'},
        {'trigger': 'level2_complete', 'source': 'level2', 'dest': 'level3'},
        {'trigger': 'player_exit', 'source': 'level3', 'dest': 'idle'},
        {'trigger': 'level3_complete', 'source': 'level3', 'dest': 'outro'},
        {'trigger': 'player_exit', 'source': 'outro', 'dest': 'idle'},
        {'trigger': 'stop', 'source': 'idle', 'dest': 'stopped'}
        ]

    def __init__(self,
                 pong_api: PongAPI):
        """
        Initializes the BaseStateMachine with the given BaseState.

        Parameters:
        pong_api (PongAPI): The PongAPI instance to observe.
        """
        self.pong_api = pong_api
        self.machine = Machine(model=self, states=BaseStateMachine.states, transitions=BaseStateMachine.transitions, initial='stopped', ignore_invalid_triggers=True)

    def start_machine(self):
        """
        Starts the BaseStateMachine by transitioning to the 'idle' state.
        """
        logging.info("BaseStateMachine Start")
        self.pong_api.register_observer('game/state', self.on_game_state)
        self.start()

        while True:
            # logging.info("Starting BaseStateMachine loop")
            time.sleep(5)

    def on_game_state(self, message):
        """
        Callback method for when the BaseState changes.

        Parameters:
        changed_state (dict): The changed state values.
        """
        print(f"Game State Transition: {message['transition']}")

        state_transition = message['transition']
        if state_transition == 'start':
            self.start()
        elif state_transition == 'player_ready':
            self.player_ready()
        elif state_transition == 'player_exit':
            self.player_exit()
        elif state_transition == 'intro_complete':
            self.intro_complete()
        elif state_transition == 'move_left_intro_complete':
            self.move_left_intro_complete()
        elif state_transition == 'start_game':
            self.start_game()
        elif state_transition == 'level1_complete':
            self.level1_complete()
        elif state_transition == 'level2_complete':
            self.level2_complete()
        elif state_transition == 'level3_complete':
            self.level3_complete()
        elif state_transition == 'game_complete':
            self.game_complete()
        elif state_transition == 'stop':
            self.stop()
        else:
            logging.info(f"Unknown state transition: {state_transition}")
            return "Default case"

    # def _on_state_change(self, changed_state):
    #     """
    #     Callback method for when the BaseState changes.

    #     Parameters:
    #     changed_state (dict): The changed state values.
    #     """
    #     # logging.info("current_state: %s", self.base_state)
    #     logging.info("changed_state: %s", changed_state)
    #     if 'game_state_transition' in changed_state:
    #         state_transition = changed_state['game_state_transition']
    #         if state_transition == 'start':
    #             self.start()
    #         elif state_transition == 'player_ready':
    #             self.player_ready()
    #         elif state_transition == 'player_exit':
    #             self.player_exit()
    #         elif state_transition == 'intro_complete':
    #             self.intro_complete()
    #         elif state_transition == 'move_left_intro_complete':
    #             self.move_left_intro_complete()
    #         elif state_transition == 'start_game':
    #             self.start_game()
    #         elif state_transition == 'level1_complete':
    #             self.level1_complete()
    #         elif state_transition == 'level2_complete':
    #             self.level2_complete()
    #         elif state_transition == 'level3_complete':
    #             self.level3_complete()
    #         elif state_transition == 'game_complete':
    #             self.game_complete()
    #         elif state_transition == 'stop':
    #             self.stop()
    #         else:
    #             logging.info(f"Unknown state transition: {state_transition}")
    #             return "Default case"

class DiagramModel():
    pass

if __name__ == "__main__":
    logging.info("Diagram BaseStateMachine.")
    model = DiagramModel() # type: ignore
    # base_state_machine = GraphMachine(model, states=BaseStateMachine.states, transitions=BaseStateMachine.transitions, initial='stopped', show_auto_transitions=True) # type: ignore
    base_state_machine = GraphMachine(model, states=BaseStateMachine.states, transitions=BaseStateMachine.transitions, initial='stopped') # type: ignore
    model.get_graph(title='State Machine Diagram for Pong').draw('dw_state_machine_diagram.png', prog='dot') # type: ignore

File: modules/shared/js_packages/dw/state-machine/lib/BaseStateMachine.js
/**
 * @file BaseStateMachine.js
 * @class BaseStateMachine
 * @description Implements a state machine using 'javascript-state-machine'.
 */
import StateMachine from 'javascript-state-machine';

// class BaseStateMachine {
export default class BaseStateMachine {
  constructor(pongAPI) {
    this.pongAPI = pongAPI;
    this.bottomPaddleLeft = false;
    this.fsm = new StateMachine({
      init: 'stopped',
      transitions: [
        { name: 'start', from: 'stopped', to: 'idle' },
        { name: 'playerReady', from: 'idle', to: 'intro' },
        { name: 'playerExit', from: 'intro', to: 'idle' },
        { name: 'introComplete', from: 'intro', to: 'moveIntro' },
        { name: 'playerExit', from: 'moveIntro', to: 'idle' },
        { name: 'moveIntroComplete', from: ['idle', 'intro', 'moveIntro'], to: 'level1Intro' },
        { name: 'level1IntroComplete', from: 'level1Intro', to: 'level1' },
        { name: 'playerExit', from: 'level1', to: 'idle' },
        { name: 'level1Complete', from: 'level1', to: 'level2Intro' },
        { name: 'level2IntroComplete', from: 'level2Intro', to: 'level2' },
        { name: 'playerExit', from: 'level2', to: 'idle' },
        { name: 'level2Complete', from: 'level2', to: 'level3Intro' },
        { name: 'level3IntroComplete', from: 'level3Intro', to: 'level3' },
        { name: 'playerExit', from: 'level3', to: 'idle' },
        { name: 'level3Complete', from: 'level3', to: 'outro' },
        { name: 'gameComplete', from: ['idle', 'outro'], to: 'idle' },
        { name: 'stop', from: 'idle', to: 'stopped' },
      ],
      methods: {
        onEnterStopped: this.onEnterStopped.bind(this),
        onLeaveStopped: this.onLeaveStopped.bind(this),
        onEnterIdle: this.onEnterIdle.bind(this),
        onLeaveIdle: this.onLeaveIdle.bind(this),
        onEnterIntro: this.onEnterIntro.bind(this),
        onLeaveIntro: this.onLeaveIntro.bind(this),
        onEnterMoveIntro: this.onEnterMoveIntro.bind(this),
        onLeaveMoveIntro: this.onLeaveMoveIntro.bind(this),
        onEnterLevel1Intro: this.onEnterLevel1Intro.bind(this),
        onLeaveLevel1Intro: this.onLeaveLevel1Intro.bind(this),
        onEnterLevel2Intro: this.onEnterLevel2Intro.bind(this),
        onLeaveLevel2Intro: this.onLeaveLevel2Intro.bind(this),
        onEnterLevel3Intro: this.onEnterLevel3Intro.bind(this),
        onLeaveLevel3Intro: this.onLeaveLevel3Intro.bind(this),
        onEnterLevel1: this.onEnterLevel1.bind(this),
        onLeaveLevel1: this.onLeaveLevel1.bind(this),
        onEnterLevel2: this.onEnterLevel2.bind(this),
        onLeaveLevel2: this.onLeaveLevel2.bind(this),
        onEnterLevel3: this.onEnterLevel3.bind(this),
        onLeaveLevel3: this.onLeaveLevel3.bind(this),
        onEnterOutro: this.onEnterOutro.bind(this),
        onLeaveOutro: this.onLeaveOutro.bind(this),
        onInvalidTransition: function(transition, from, to) {
          console.log("transition '%s' not allowed from state '%s'", transition, from);
        }
      },
    });
  }
  
  startMachine() {
    console.log('BaseStateMachine Entered startMachine'); 
    // this.baseState.addObserver(this, this.onStateChange.bind(this));
    this.pongAPI.registerObserver('game/state', this.onGameState.bind(this));
    this.fsm.start();
    // this.runLoop();
  }

  // /**
  //  * Continuous loop with a sleep time of 5 seconds.
  //  */
  // async runLoop() {
  //   while (true) {
  //     await new Promise(resolve => setTimeout(resolve, 5000));
  //   }
  // }

  /**
   * Callback for game/state topic.
   * @param {object} message - The message received.
   */
  onGameState(message) {
    // console.log('Game State Message:', message);
    const stateTransition = message.transition;
    switch (stateTransition) {
          case 'start':
            this.fsm.start();
              break;
          case 'player_ready':
            this.fsm.playerReady();
            break;
          case 'player_exit':
            this.fsm.playerExit();
            break;
          case 'intro_complete':
            this.fsm.introComplete();
            break;
          case 'move_intro_complete':
            this.fsm.moveIntroComplete();
            break;
          case 'level1_intro_complete':
            this.fsm.level1IntroComplete();
            break;
          case 'level1_complete':
            this.fsm.level1Complete();
            break;
          case 'level2_intro_complete':
            this.fsm.level2IntroComplete();
            break;
          case 'level2_complete':
            this.fsm.level2Complete();
            break;
          case 'level3_intro_complete':
            this.fsm.level3IntroComplete();
            break;
          case 'level3_complete':
            this.fsm.level3Complete();
            break;
          case 'game_complete':
              this.fsm.gameComplete();
              break;
          case 'stop':
              this.fsm.stop();
              break;
          default:
              console.log("Unknown state transition:", stateTransition);
      // }    
  }
  }
  
  // Define onEnter<state> and onLeave<state> methods for each state
  onEnterStopped() { }
  onLeaveStopped() { }
  onEnterIdle() { 
    this.bottomPaddleLeft = false; 
  }
  onLeaveIdle() { }
  onEnterIntro() { }
  onLeaveIntro() { }
  onEnterMoveIntro() { }
  onLeaveMoveIntro() { }
  onEnterLevel1Intro() { }
  onLeaveLevel1Intro() { }
  onEnterLevel2Intro() { }
  onLeaveLevel2Intro() { }
  onEnterLevel3Intro() { }
  onLeaveLevel3Intro() { }
  onEnterLevel1() { }
  onLeaveLevel1() { }
  onEnterLevel2() { }
  onLeaveLevel2() { }
  onEnterLevel3() { }
  onLeaveLevel3() { }
  onEnterOutro() { }
  onLeaveOutro() { }
 
}